ops/charm.py

```diff
@@ -18,8 +18,12 @@ def restore(self, snapshot):
         if event_action_name != env_action_name:
             # This could only happen if the dev manually emits the action, or from a bug.
             raise RuntimeError('action event kind does not match current action')
-        # Params are loaded at restore rather than __init__ because the model is not available in __init__.
-        self.params = self.framework.model._backend.action_get()
+
+    @property
+    def params(self):
+        # NoneType will be returned if an action has no parameters defined so getattr is used instead of a check for NoneType.
+        self._params = getattr(self, '_params', self.framework.model._backend.action_get())
+        return self._params
```

Comment:
This syntax causes us to still evaluate _backend.action_get() on every lookup to self.params. Because functions passed to a function must be evaluated before the function is called.
What you want is something more like:
```
def __init__(self):
...
  self._params = None
...
def params(self):
  if self._params is None:
    self._params = self.framework.model._backend.action_get()
  return self._params
```

Also, this doesn't seem related to the rest of the patch, probably better to put it into its own patch.

Reply:
The problem with ```if self._params is None:``` is that action-get will return `None` if there are no parameters defined. Thus on any `.params` access there will always be a hook tool execution.

I don't think this is terrible because people wouldn't normally access `params` if they know that an action does not accept any.